nupm/registry.nu

```diff
@@ -0,0 +1,252 @@
+                    # to avoid confusion with custom|script|module type enumberable
+                    source: $pkg.type,
+                    version: $pkg.version,
+                    # description: ($pkg.description? | default "")
+                }
+            }
+        } | flatten
+    } catch {|err|
+        throw-error $"Failed to fetch registry data from '($registry_url)': ($err.msg)"
+    }
+}
+
+# Add a new registry
+@example "Add a new registry" { nupm registry add my-registry https://example.com/registry.nuon }
+export def --env add [
+    name: string,       # Name of the registry
+    url: string,        # URL or path to the registry
+    --save,             # Whether to commit the change to the registry index
+] {
+    if ($name in $env.NUPM_REGISTRIES) {
+        throw-error $"Registry '($name)' already exists. Use 'nupm registry update' to modify it."
+    }
+    $env.NUPM_REGISTRIES = $env.NUPM_REGISTRIES | insert $name $url
+
+    if $save {
+      $env.NUPM_REGISTRIES | save --force $env.NUPM_INDEX_PATH
+    }
+
+    print $"Registry '($name)' added successfully."
+}
+
+# Remove a registry
+@example "Remove a registry" { nupm registry remove my-registry }
+export def --env remove [
+    name: string        # Name of the registry to remove
+    --save,             # Whether to commit the change to the registry index
+] {
+    $env.NUPM_REGISTRIES = $env.NUPM_REGISTRIES | reject $name
+
+    if $save {
+      $env.NUPM_REGISTRIES | save --force $env.NUPM_INDEX_PATH
+    }
+
+    print $"Registry '($name)' removed successfully."
+}
+
+# Update a given registry url
+@example "Update registry URL" { nupm registry set-url my-registry https://new-url.com/registry.nuon }
+export def --env set-url [
+    name: string,   # Name of the registry to update
+    url: string,
+    --save,         # Whether to commit the change to the registry index
+]: nothing -> nothing {
+    $env.NUPM_REGISTRIES = $env.NUPM_REGISTRIES | update $name $url
+
+    if $save {
+      $env.NUPM_REGISTRIES | save --force $env.NUPM_INDEX_PATH
+    }
+
+    print $"Registry '($name)' URL updated successfully."
+}
+
+# https://www.nushell.sh/book/configuration.html#macos-keeping-usr-bin-open-as-open
+alias nu-rename = rename
+# Rename a registry
+@example "Rename a registry" { nupm registry rename my-registry our-registry }
+export def --env rename [
+    name: string,   # Name of the registry to update
+    new_name: string,
+    --save,         # Whether to commit the change to the registry index
+] {
+    $env.NUPM_REGISTRIES = $env.NUPM_REGISTRIES | nu-rename --column { $name: $new_name }
+
+    if $save {
+      $env.NUPM_REGISTRIES | save --force $env.NUPM_INDEX_PATH
+    }
+
+    print $"Registry '($name)' renamed successfully."
+}
+
+# Fetch and cache registry data locally
+@example "Fetch a specific registry" { nupm registry fetch nupm }
+@example "Fetch all registries" { nupm registry fetch --all }
+export def fetch [
```

Comment:
So this fetches `registry.nuon` and `<package>.nuon` files of the desired registries into cache? Why do we need to cache registries? It introduces a problem if the original registry is updated, how does nupm know to update the cache? We should cache downloaded packages with proper hashing etc., but looking up a registry is just fetching a single file, it shouldn't be a problem and attempting to cache it would introduce unnecessary problems IMO.